ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterProjectTransposeRule.java

```diff
@@ -339,6 +339,11 @@ private void check(Filter filter) {
       final RexNode filterCondition = simplify.simplify(filter.getCondition());
 
       final Set<Integer> inputRefs = HiveCalciteUtil.getInputRefs(newCondition);
+      // if the new IS NOT NULL has no input ref, there is redundancy here, bail out
```

Comment:
the above code blindly states that the top level call is an `IS NOT NULL` without even checking it
can we be sure in that? because it seems like it removes the top level call without even a check!

Reply:
This auxiliary method is only called in the context of this method, which only targets `IS NOT NULL` calls: [HiveFilterProjectTransposeRule.java#L265](https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveFilterProjectTransposeRule.java#L265)

The `check` method does not remove any call by itself, it is just in charge to check if the new `IS NOT NULL` that we want to create would not be redundant, if so it prevents the push by making the rule bail out.

I agree that the name of the method is not ideal, maybe we can rename it as `checkRedundancyOfIsNotNull` or something along that line?